Composer/packages/server/src/externalContentProvider/azureBotServiceProvider.ts

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+import { PublishTarget } from '@bfc/shared';
+import * as axios from 'axios';
+import jwtDecode from 'jwt-decode';
+
+import { authService } from '../services/auth/auth';
+
+import { IContentProviderMetadata, ExternalContentProvider, BotContentInfo } from './externalContentProvider';
+
+export type AzureBotServiceMetadata = IContentProviderMetadata & {
+  resourceId: string;
+  botName: string;
+  appId?: string;
+  appPasswordHint?: string;
+  subscriptionId?: string;
+  resourceGroup?: string;
+  armEndpoint?: string;
+  tenantId: string;
+};
+
+export class AzureBotServiceProvider extends ExternalContentProvider<AzureBotServiceMetadata> {
+  constructor(metadata: AzureBotServiceMetadata) {
+    super(metadata);
+  }
+  public async downloadBotContent(): Promise<BotContentInfo> {
+    throw new Error('ABS Not support');
+  }
+  public async cleanUp(): Promise<void> {
+    throw new Error('Method not implemented.');
+  }
+  public async getAlias(): Promise<string> {
+    const alias = `abs-${this.metadata.botName}-${this.metadata.appId}`;
+    return alias;
+  }
+
+  public async authenticate(): Promise<string> {
+    return this.getVaultAccessToken();
+  }
+
+  public async generateProfile(): Promise<PublishTarget> {
+    const appId = this.metadata.appId;
+    // parse subscriptionId ... from this.metadata
+    const temp = { ...this.metadata };
+    const subs = this.metadata.resourceId.match(/subscriptions\/([\w-]*)\//);
+    const groups = this.metadata.resourceId.match(/resourceGroups\/([^/]*)/);
+    const names = this.metadata.resourceId.match(/botServices\/([^/]*)/);
+    temp.subscriptionId = (subs && subs.length > 0 && subs[1]) || '';
+    temp.resourceGroup = (groups && groups.length > 0 && groups[1]) || '';
+    temp.botName = (names && names.length > 0 && names[1]) || '';
+    delete temp.appId;
+
+    // transform key vault token hint to microsoft app password
+    const hint = this.metadata.appPasswordHint || '';
+    delete temp.appPasswordHint;
+    const vaultNames = hint.match(/vaults\/([^/]*)/);
+    const secretNames = hint.match(/secrets\/([^/]*)/);
+    const resourceGroupNames = hint.match(/resourceGroups\/([^/]*)/);
+    const vaultName = (vaultNames && vaultNames.length > 0 && vaultNames[1]) || '';
+    const secretName = (secretNames && secretNames.length > 0 && secretNames[1]) || '';
+    const resourceGroupName = (resourceGroupNames && resourceGroupNames.length > 0 && resourceGroupNames[1]) || '';
+    const tenantId = this.metadata.tenantId;
+
+    // get token
+    const armToken = await this.getArmAccessToken(tenantId);

```

Comment:
There might be some exceptions here:
1. get token failure
2. SetPolicy failure
3. get secrets failure

we should add logic to handle those exceptions.